File: DID.Native.SDK/src/utility/HDkey.c
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <openssl/opensslv.h>
#include <openssl/rand.h>
#include <openssl/ec.h>
#include <openssl/obj_mac.h>

#include "HDkey.h"
#include "BRBIP39Mnemonic.h"
#include "BRBIP39WordsEn.h"
#include "BRBIP39WordsChs.h"
#include "BRBIP32Sequence.h"
#include "BRCrypto.h"
#include "BRBase58.h"
#include "BRInt.h"

static unsigned char PADDING_IDENTITY = 0x67;
static unsigned char PADDING_STANDARD = 0xAD;

static const char **get_word_list(int language)
{
    switch (language) {
    case LANGUAGE_ENGLISH:
        return BRBIP39WordsEn;

    case LANGUAGE_CHINESE_SIMPLIFIED:
        return BRBIP39WordsChs;

    default:
        return NULL;
    }
}

// need to free after use this function
const char *HDkey_GenerateMnemonic(int language)
{
    unsigned char rand[16];
    const char **word_list;
    char *phrase;
    size_t len;

    //create random num
    if (RAND_bytes(rand, sizeof(rand)) != 1)
        return NULL;

    word_list = get_word_list(language);
    if (!word_list)
        return NULL;

    len = BRBIP39Encode(NULL, 0, word_list, rand, sizeof(rand));

    phrase = (char *)malloc(len);
    if (!phrase)
        return NULL;

    BRBIP39Encode(phrase, len, word_list, rand, sizeof(rand));
    return (const char *)phrase;
}

uint8_t *HDkey_GetSeedFromMnemonic(const char *mmemonic,
        const char* passphrase, int language, uint8_t *seed)
{
    int len;
    const char **word_list;

    if (!mmemonic || !seed)
        return NULL;

    word_list = get_word_list(language);
    if (!word_list)
        return NULL;

    if (!BRBIP39PhraseIsValid(word_list, mmemonic))
        return NULL;

    BRBIP39DeriveKey((UInt512 *)seed, mmemonic, passphrase);

    return seed;
}

MasterPublicKey* HDkey_GetMasterPublicKey(const uint8_t *seed, int coinType,
    MasterPublicKey *masterkey)
{
    char publicKey[33];
    UInt256 chainCode;
    BRKey key;

    if (!seed || !masterkey)
        return NULL;

    BRBIP32PrivKeyPath(&key, &chainCode, (const void *)seed, SEED_BYTES,
            3, 44 | BIP32_HARD, coinType | BIP32_HARD, 0 | BIP32_HARD);

    getPubKeyFromPrivKey(publicKey, &(key.secret));

    masterkey->fingerPrint = BRKeyHash160(&key).u32[0];
    memcpy(masterkey->chainCode, (uint8_t*)&chainCode, sizeof(chainCode));
    memcpy(masterkey->publicKey, publicKey, sizeof(publicKey));

    var_clean(&chainCode);

    return masterkey;
}

uint8_t *HDkey_GetSubPrivateKey(const uint8_t *seed, int coinType, int chain,
        int index, uint8_t *privatekey)
{
    UInt256 chainCode;
    BRKey key;

    if (!seed || !privatekey)
        return NULL;

    BRBIP32PrivKeyPath(&key, &chainCode, (const void *)seed, SEED_BYTES,
            5, 44 | BIP32_HARD, coinType | BIP32_HARD,
            0 | BIP32_HARD, chain, index);
    var_clean(&chainCode);

    assert(sizeof(key.secret.u8) == PRIVATEKEY_BYTES);

    memcpy(privatekey, key.secret.u8, PRIVATEKEY_BYTES);
    return privatekey;
}

uint8_t *HDkey_GetSubPublicKey(MasterPublicKey* masterkey, int chain,
        int index, uint8_t *publickey)
{
    if (!masterkey || !publickey)
        return NULL;

    BRMasterPubKey brPublicKey;
    brPublicKey.fingerPrint = masterkey->fingerPrint;
    memcpy((uint8_t*)&brPublicKey.chainCode, masterkey->chainCode,
            sizeof(brPublicKey.chainCode));
    memcpy(brPublicKey.pubKey, masterkey->publicKey, sizeof(brPublicKey.pubKey));

    assert(BRBIP32PubKey(NULL, 0, brPublicKey, chain, index) == PUBLICKEY_BYTES);

    BRBIP32PubKey(publickey, PUBLICKEY_BYTES, brPublicKey, chain, index);
    return publickey;
}

char *HDkey_GetIdString(unsigned char *publickey, char *address, size_t len)
{
    unsigned char redeem_script[35];
    unsigned int md32[32];
    unsigned char md20[20];
    unsigned char program_hash[21];
    unsigned char bin_address[25];
    size_t expected_len;

    if (!publickey || !address || !len)
        return NULL;

    redeem_script[0] = 33;
    memcpy(redeem_script + 1, publickey, 33);
    redeem_script[34] = PADDING_STANDARD;
    BRHash160(md20, redeem_script, sizeof(redeem_script));

    program_hash[0] = PADDING_IDENTITY;
    memcpy(program_hash + 1, md20, sizeof(md20));

    BRSHA256_2(md32, program_hash, sizeof(program_hash));

    memcpy(bin_address, program_hash, sizeof(program_hash));
    memcpy(bin_address + sizeof(program_hash), md32, 4);

    expected_len = BRBase58Encode(NULL, 0, bin_address, sizeof(bin_address));
    if (len < expected_len)
        return NULL;

    BRBase58Encode(address, len, bin_address, sizeof(bin_address));
    return address;
}

File: DID.Java.SDK/src/main/java/org/elastos/did/crypto/HDKey.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did.crypto;

import static org.bitcoinj.core.ECKey.CURVE;
import static org.bitcoinj.core.ECKey.CURVE_PARAMS;

import java.math.BigInteger;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.ECParameterSpec;
import java.security.spec.ECPrivateKeySpec;
import java.security.spec.ECPublicKeySpec;
import java.security.spec.InvalidKeySpecException;

import org.bitcoinj.core.ECKey.ECDSASignature;
import org.bitcoinj.core.Sha256Hash;
import org.bitcoinj.core.SignatureDecodeException;
import org.bitcoinj.crypto.ChildNumber;
import org.bitcoinj.crypto.DeterministicKey;
import org.bitcoinj.crypto.HDKeyDerivation;
import org.bitcoinj.crypto.HDPath;
import org.bitcoinj.params.MainNetParams;
import org.elastos.did.Mnemonic;
import org.elastos.did.exception.InvalidKeyException;
import org.spongycastle.crypto.digests.RIPEMD160Digest;
import org.spongycastle.crypto.digests.SHA256Digest;
import org.spongycastle.crypto.params.ECPublicKeyParameters;
import org.spongycastle.jce.spec.ECNamedCurveSpec;

public class HDKey {
	public static final int PUBLICKEY_BYTES = 33;
	public static final int PRIVATEKEY_BYTES = 32;
	public static final int SEED_BYTES = 64;
	public static final int EXTENDED_KEY_BYTES = 82;
	public static final int EXTENDED_PRIVATEKEY_BYTES = EXTENDED_KEY_BYTES;
	public static final int EXTENDED_PUBLICKEY_BYTES = EXTENDED_KEY_BYTES;


    private final static byte PADDING_IDENTITY	= 0x67;
    private final static byte PADDING_STANDARD	= (byte)0xAD;

    private DeterministicKey key;

	// Derive path: m/44'/0'/0'/0/index
	public static final String DERIVE_PATH_PREFIX = "44H/0H/0H/0/";

	// Pre-derive publickey path: m/44'/0'/0'
	public static final String PRE_DERIVED_PUBLICKEY_PATH = "44H/0H/0H";

	private HDKey(DeterministicKey key) {
		this.key = key;
	}

	public HDKey(String mnemonic, String passphrase) {
		this(Mnemonic.toSeed(mnemonic, passphrase));
	}

	public HDKey(byte[] seed) {
		this(HDKeyDerivation.createMasterPrivateKey(seed));
	}

	public byte[] getPrivateKeyBytes() {
		return key.getPrivKeyBytes();
	}

	public String getPrivateKeyBase58() {
		return Base58.encode(getPrivateKeyBytes());
	}

	public byte[] getPublicKeyBytes() {
		return key.getPubKey();
	}

	public String getPublicKeyBase58() {
		return Base58.encode(getPublicKeyBytes());
	}

	public byte[] serialize() {
		return Base58.decode(serializeBase58());
	}

	public String serializeBase58() {
		return key.serializePrivB58(MainNetParams.get());
	}

	public byte[] serializePublicKey() {
		return Base58.decode(serializePublicKeyBase58());
	}

    public String serializePublicKeyBase58() {
    	return key.serializePubB58(MainNetParams.get());
    }

	public static HDKey deserialize(byte[] keyData) {
		/*
		DeterministicKey k = DeterministicKey.deserialize(
				MainNetParams.get(), keyData);
		return new HDKey(k);
		*/
		return deserializeBase58(Base58.encode(keyData));
	}

	public static HDKey deserializeBase58(String keyData) {
		DeterministicKey k = DeterministicKey.deserializeB58(
				keyData, MainNetParams.get());
		return new HDKey(k);
	}

	public static byte[] paddingToExtendedPrivateKey(byte[] privateKeyBytes) {
		byte[] extendedPrivateKeyBytes = new byte[EXTENDED_PRIVATEKEY_BYTES];

		int version = MainNetParams.get().getBip32HeaderP2PKHpriv();
		extendedPrivateKeyBytes[0] = (byte)((version >> 24) & 0xFF);
		extendedPrivateKeyBytes[1] = (byte)((version >> 16) & 0xFF);
		extendedPrivateKeyBytes[2] = (byte)((version >> 8) & 0xFF);
		extendedPrivateKeyBytes[3] = (byte)(version & 0xFF);

		System.arraycopy(privateKeyBytes, 0,
				extendedPrivateKeyBytes, 46, 32);

		byte[] hash = Sha256Hash.hashTwice(extendedPrivateKeyBytes, 0, 78);
		System.arraycopy(hash, 0, extendedPrivateKeyBytes, 78, 4);

		return extendedPrivateKeyBytes;
	}

	public static byte[] paddingToExtendedPublicKey(byte[] publicKeyBytes) {
		byte[] extendedPublicKeyBytes = new byte[EXTENDED_PUBLICKEY_BYTES];

		int version = MainNetParams.get().getBip32HeaderP2PKHpub();
		extendedPublicKeyBytes[0] = (byte)((version >> 24) & 0xFF);
		extendedPublicKeyBytes[1] = (byte)((version >> 16) & 0xFF);
		extendedPublicKeyBytes[2] = (byte)((version >> 8) & 0xFF);
		extendedPublicKeyBytes[3] = (byte)(version & 0xFF);

		System.arraycopy(publicKeyBytes, 0,
				extendedPublicKeyBytes, 45, 33);

		byte[] hash = Sha256Hash.hashTwice(extendedPublicKeyBytes, 0, 78);
		System.arraycopy(hash, 0, extendedPublicKeyBytes, 78, 4);

		return extendedPublicKeyBytes;
	}

	public HDKey derive(String path) {
		HDPath derivePath = HDPath.parsePath(path);

		DeterministicKey child = key;
		for (ChildNumber childNumber: derivePath)
			child = HDKeyDerivation.deriveChildKey(child, childNumber);

		return new HDKey(child);
	}

	public HDKey derive(int index, boolean hardened) {
		ChildNumber childNumber = new ChildNumber(index, hardened);
		return new HDKey(HDKeyDerivation.deriveChildKey(key, childNumber));
	}

	public HDKey derive(int index) {
		return derive(index, false);
	}

    public KeyPair getJCEKeyPair() throws InvalidKeyException {
    	ECParameterSpec paramSpec = new ECNamedCurveSpec(
        		"secp256r1", CURVE_PARAMS.getCurve(), CURVE_PARAMS.getG(),
        		CURVE_PARAMS.getN(), CURVE_PARAMS.getH());

        KeyFactory keyFactory = null;
		try {
			keyFactory = KeyFactory.getInstance("EC");
		} catch (NoSuchAlgorithmException ignore) {
			// never happen
		}

    	PublicKey pub = null;
    	PrivateKey priv = null;

    	try {
    		ECPublicKeyParameters pubParams = new ECPublicKeyParameters(
    				CURVE_PARAMS.getCurve().decodePoint(getPublicKeyBytes()), CURVE);
    		ECPublicKeySpec pubSpec = new ECPublicKeySpec(new java.security.spec.ECPoint(
    				pubParams.getQ().getXCoord().toBigInteger(),
    				pubParams.getQ().getYCoord().toBigInteger()), paramSpec);
    		pub = keyFactory.generatePublic(pubSpec);

	    	if (key.hasPrivKey()) {
	    		BigInteger keyInt = new BigInteger(1, getPrivateKeyBytes());
	    		ECPrivateKeySpec privSpec = new ECPrivateKeySpec(keyInt, paramSpec);
	    		priv = keyFactory.generatePrivate(privSpec);
	    	}
		} catch (InvalidKeySpecException e) {
			throw new InvalidKeyException();
		}
    	return new KeyPair(pub, priv);
    }

	private static byte[] getRedeemScript(byte[] pk) {
        byte[] script = new byte[35];
        script[0] = 33;
        System.arraycopy(pk, 0, script, 1, 33);
        script[34] = PADDING_STANDARD;
        return script;
	}

	private static byte[] sha256Ripemd160(byte[] input) {
        byte[] sha256 = new byte[32];

        SHA256Digest sha256Digest = new SHA256Digest();
        sha256Digest.update(input, 0, input.length);
        sha256Digest.doFinal(sha256, 0);

        RIPEMD160Digest digest = new RIPEMD160Digest();
        digest.update(sha256, 0, sha256.length);
        byte[] out = new byte[20];
        digest.doFinal(out, 0);
        return out;
    }

	private static byte[] getBinAddress(byte[] pk) {
        byte[] script = getRedeemScript(pk);

        byte[] hash = sha256Ripemd160(script);
        byte[] programHash = new byte[hash.length + 1];
        programHash[0] = PADDING_IDENTITY;
        System.arraycopy(hash, 0, programHash, 1, hash.length);

        hash = Sha256Hash.hashTwice(programHash);
        byte[] binAddress = new byte[programHash.length+4];
        System.arraycopy(programHash, 0, binAddress, 0, programHash.length);
        System.arraycopy(hash, 0, binAddress, programHash.length, 4);

        return binAddress;
	}

	public byte[] getBinAddress() {
		return getBinAddress(getPublicKeyBytes());
	}

	public String getAddress() {
        return Base58.encode(getBinAddress());
	}

	public static String toAddress(byte[] pk) {
		return Base58.encode(getBinAddress(pk));
	}

	public static boolean isAddressValid(String address) {
		byte[] binAddress = Base58.decode(address);

		if (binAddress.length != 25)
			return false;

		if (binAddress[0] != PADDING_IDENTITY)
			return false;

		byte[] hash = Sha256Hash.hashTwice(binAddress, 0, 21);

		return (hash[0] == binAddress[21] && hash[1] == binAddress[22]
				&& hash[2] == binAddress[23] && hash[3] == binAddress[24]);
	}

	public byte[] sign(byte[] sha256Hash) {
		ECDSASignature sig = key.sign(Sha256Hash.wrap(sha256Hash));
		return sig.encodeToDER();
	}

	public byte[] signData(byte[] ... inputs) {
		byte[] hash = sha256Digest(inputs);

		return sign(hash);
	}

	public boolean verify(byte[] sha256Hash, byte[] signature) {
		try {
			return key.verify(sha256Hash, signature);
		} catch (SignatureDecodeException e) {
			return false;
		}
	}

	public boolean verifyData(byte[] signature, byte[] ... inputs) {
		byte[] hash = sha256Digest(inputs);

		return verify(hash, signature);
	}

	private static byte[] sha256Digest(byte[] ... inputs) {
		byte digest[] = new byte[32];

		SHA256Digest sha256 = new SHA256Digest();

		for (byte[] input : inputs)
			sha256.update(input, 0, input.length);

		sha256.doFinal(digest, 0);

		return digest;
	}

	public void wipe() {
		// TODO:
	}
}
